lib/mix/tasks/elixir_make.checksum.ex

```diff
@@ -103,38 +103,79 @@ defmodule Mix.Tasks.ElixirMake.Checksum do
     tasks =
       Task.async_stream(
         urls,
-        fn {{_target, _nif_version}, url} -> {url, Artefact.download(url)} end,
+        fn {{_target, _nif_version}, url} ->
+          checksum_algo = Artefact.checksum_algo()
+          checksum_file_url = "#{url}.#{Atom.to_string(checksum_algo)}"
+          artifact_checksum = Artefact.download(checksum_file_url)
+
+          from_checksum_file =
+            case artifact_checksum do
+              {:ok, body} ->
+                case String.split(body, " ", trim: true) do
+                  [checksum, basename] ->
+                    {:ok,
+                     {:checksum, url,
+                      %Artefact{
+                        basename: basename,
+                        checksum: checksum,
+                        checksum_algo: checksum_algo
+                      }}}
+
+                  _ ->
+                    :download_and_compute
+                end
+
+              _ ->
+                :download_and_compute
+            end
+
+          case from_checksum_file do
+            {:ok, packed} ->
+              packed
+
+            _ ->
+              {:download, url, Artefact.download(url)}
+          end
+        end,
         timeout: :infinity,
         ordered: false
       )
 
     cache_dir = Artefact.cache_dir()
 
-    Enum.flat_map(tasks, fn {:ok, {url, download}} ->
-      case download do
-        {:ok, body} ->
-          basename = basename_from_url(url)
-          path = Path.join(cache_dir, basename)
-          File.write!(path, body)
-          artefact = Artefact.checksum(basename, body)
-
-          Mix.shell().info(
-            "NIF cached at #{path} with checksum #{artefact.checksum} (#{artefact.checksum_algo})"
-          )
-
-          [artefact]
-
-        result ->
-          if ignore_unavailable? do
-            msg = "Skipped unavailable NIF artifact. Reason: #{inspect(result)}"
-            Mix.shell().info(msg)
-          else
-            msg = "Could not finish the download of NIF artifacts. Reason: #{inspect(result)}"
-            Mix.shell().error(msg)
-          end
+    Enum.flat_map(tasks, fn
+      {:ok, :checksum, _url, artefact} ->
```

Comment:
I don't think this will ever match:

```suggestion
      {:ok, {:checksum, _url, artefact}} ->
```